runtime/android/core/src/org/xwalk/core/XWalkLibraryLoader.java

```diff
@@ -0,0 +1,305 @@
+            Log.d(TAG, "DecompressionTask cancelled");
+            sActiveTask = null;
+            mListener.onDecompressionCancelled();
+        }
+
+        @Override
+        protected void onPostExecute(Integer result) {
+            Log.d(TAG, "DecompressionTask finished");
+            sActiveTask = null;
+            mListener.onDecompressionCompleted();
+        }
+    }
+
+    private static class DownloadTask extends AsyncTask<Void, Integer, Integer> {
+        private static final int QUERY_INTERVAL_MS = 100;
+        private static final int MAX_PAUSED_COUNT = 6000; // 10 minutes
+        private static final int MAX_RUNNING_COUNT = 18000; // 30 minutes
+
+        private DownloadListener mListener;
+        private String mDownloadUrl;
+        private DownloadManager mDownloadManager;
+        private long mDownloadId;
+
+        DownloadTask(DownloadListener listener, Context context, String url) {
+            super();
+            mListener = listener;
+            mDownloadUrl = url;
+            mDownloadManager= (DownloadManager) context.getSystemService(Context.DOWNLOAD_SERVICE);
+        }
+
+        @Override
+        protected void onPreExecute() {
+            Log.d(TAG, "DownloadTask started, " + mDownloadUrl);
+            sActiveTask = this;
+
+            Request request = new Request(Uri.parse(mDownloadUrl));
+            request.setDestinationInExternalPublicDir(
+                    Environment.DIRECTORY_DOWNLOADS, XWALK_APK_NAME);
+            mDownloadId = mDownloadManager.enqueue(request);
+
+            mListener.onDownloadStarted();
+        }
+
+        @Override
+        protected Integer doInBackground(Void... params) {
+            Query query = new Query().setFilterById(mDownloadId);
+            int pausedCount = 0;
+            int runningCount = 0;
+
+            while (!isCancelled()) {
+                try {
+                    Thread.sleep(QUERY_INTERVAL_MS);
+                } catch (InterruptedException e) {
+                    break;
+                }
+
+                Cursor cursor = mDownloadManager.query(query);
+                if (cursor == null || !cursor.moveToFirst()) continue;
+
+                int totalIdx = cursor.getColumnIndex(DownloadManager.COLUMN_TOTAL_SIZE_BYTES);
+                int downloadIdx = cursor.getColumnIndex(
+                        DownloadManager.COLUMN_BYTES_DOWNLOADED_SO_FAR);
+                int totalSize = cursor.getInt(totalIdx);
+                int downloadSize = cursor.getInt(downloadIdx);
+                if (totalSize > 0) publishProgress(downloadSize, totalSize);
+
+                int statusIdx = cursor.getColumnIndex(DownloadManager.COLUMN_STATUS);
+                int status = cursor.getInt(statusIdx);
+                if (status == DownloadManager.STATUS_FAILED ||
+                        status == DownloadManager.STATUS_SUCCESSFUL) {
+                    return status;
+                } else if (status == DownloadManager.STATUS_PAUSED) {
```

Comment:
Do we need these xxxCount?
I think this could only incur unpredicated failure, could remove this, user will cancel the downloading if they thought it's too long.